### ISSUE TITLE
[BUG] Problem loading the configuration file after manual installation

### ISSUE THREAD

**Issue by simoncarne** [Izheil/Quantum-Nox-Firefox-Dark-Full-Theme#106]
I successfully installed the MultiRow feature (using the manual method described at https://github.com/Izheil/Quantum-Nox-Firefox-Dark-Full-Theme/tree/master/Multirow%20and%20other%20functions/JS%20Loader). But, after a few days, it stopped working. 

More specifically, when I opened Firefox, I received a message "Failed to read the configuration file. Please contact your system administrator." 
![image](https://user-images.githubusercontent.com/33954467/104560890-7375e780-563e-11eb-97e5-725aeb4c4fe6.png)
 
I re-installed the MultiRow feature, exactly as before, and all was well for a while and then the same problem recurred. On this occasion, I re-installed only the config file and, once again, it is working, so the problem seems to be associated only with that file. 

I am running MultiRow on two separate PCs and I have exactly the same problem on both devices, which seems to rule out a problem on my own device. 

I am not able to discern what triggers the problem. 

I am running: 
- Windows 10 
- Firefox 84.0.2 (64 Bit) on both devices.
- The most recent version of Quantum-Nox-Firefox-Dark-Full-Theme (ie from 7 Jan 2021)

**Comment by Izheil** [comment]
When you say that you replaced the configuration file, I assume you mean config.js, right?

That problem should only happen when it can't find the configuration file specified by `defaults/pref/` files (in our case config.js).

Can you confirm that when the problem happens, config.js has read access by your user?

**Comment by simoncarne** [comment]
Thanks for picking this up so quickly.

Yes, I am referring to config.js. (Apologies for not being clear.)

Please could you explain how I can check the "read access" status. This is not a term I am familiar with. 

Yes, the only files in faults/pref/ are channel-prefs.js and config-prefs.js.

**Comment by Izheil** [comment]
Right click `config.js` -> Properties -> Security tab -> Click on your username/group.
You should see the permissions your user or user group has clicking on it.
Since we start Firefox as non-admin, check if the "read" permission is ticked on the "users" group.

**Comment by simoncarne** [comment]
Yes, config.js has read access by my User.

![Screenshot 2021-01-14 085817](https://user-images.githubusercontent.com/33954467/104567528-cfdd0500-5646-11eb-8d28-6e3ae76a2ddb.png)

**Comment by Izheil** [comment]
Just to make sure, you placed `config.js` in the same folder where `firefox.exe` is, didnt change it's name, and it's contents are the same as [this one](https://github.com/Izheil/Quantum-Nox-Firefox-Dark-Full-Theme/blob/master/Multirow%20and%20other%20functions/JS%20Loader/root/config.js)? (open config.js with notepad, opening it with word would add unnecesary lines that would make it fail loading).

**Comment by simoncarne** [comment]
Yes to all of the above. 

BUT, just so we are clear, although the config.js file I loaded was exactly as you set out, the file has acquired a comment (beginning // kl_config) and an extra line starting with lockPref.  

May I also remind you that, when I first installed/re-installed the MultiRow features, everything works for a while. The problem isn't there from the start.

**Comment by Izheil** [comment]
`Config.js` shouldn't acquire any comment at all. 
A quick search seems to point that //kl_config added comment might be related to kapersky.
That `lockPref` line prevents the preference it states from being changed on Firefox (so if it locks any of the preferences that we require to change with `config-prefs.js` for config.js to be loaded, it would fail loading).

It makes sense that it starts fine the first time (before Kapersky checks the file and then changes it, but are not loaded until the next restart), but not the next loads (after the file has already been changed, and the changes take place on Firefox).

If you indeed use kapersky, try to add an exclusion for `config.js` and `config-prefs.js` on it, replace the config.js file again, set them as read-only (right click the file -> properties -> tick the "read only" checkbox) and try restarting Firefox twice to see if it gets changed again.

Kapersky also seems to add [some .cfg file on Firefox root folder](https://bugzilla.mozilla.org/show_bug.cgi?id=1591691#c13) as well that might create issues... so you might have to set that to false like in that bugzilla (if it exists for you).

**Comment by simoncarne** [comment]
Many thanks. 

Sadly, that doesn't seem to be resolving the problem. In fact, the change in the `config.js` file may not be the issue (see Update 2 below). 

**Update 1** 
I'm not using Kapesrsky, but I am using ZoneAlarm Extreme Security. I tried adding an exception as below, but that hasn't stopped the `config,js` file changing. BUT this may not be the issue - see next update. 

![Zone alarm](https://user-images.githubusercontent.com/33954467/104582779-39b1da80-5658-11eb-97f2-866610f09be9.png)

**Update 2**
I re-installed the clean `config.js` file. Re-started Firefox. The `config.js` file has changed (in the same way as before). But the MultiRow function is still working. 

Based on past behaviour, it could be several hours (or days) before MultiRow stops working. I haven't, so far, been able to work out what triggers it.

**Comment by Izheil** [comment]
What lines exactly get added to `config.js`? It might not seem like it's the culprit, but it shouldn't get changed at all. 

Also, if it gets fixed when you replace it, the issue indeed seems to be some of the changes that the file has, even if they get loaded later on (probably because Firefox still loads the file first as is, and then loads it from cache for a while until it either gets it's file identifier changed (when it's overwritten), or when the cache expires).

It's either the antivirus modifying it, or something else... so try running a virus scan just in case.

Do you use ZoneAlarm in both of the PC's where you tested? I haven't had this kind of issues with PC's that were using either Windows Defender or Avast.

As a test, try temporally disabling ZoneAlarm, then replacing config.js again, then firing Firefox to see if the change happens.

**Comment by simoncarne** [comment]
I tried the test you suggested (disable ZoneAlarm, replace `config.js`, re-start Firefox). The `config.js`  file changed again, in exactly the same way. The file reads as follows:
![Config](https://user-images.githubusercontent.com/33954467/104586846-f5293d80-565d-11eb-95dd-51c6541c5f50.png)

I have also restarted ZoneAlarm and scanned both versions of `config.js` and, in both cases, ZoneAlarm reports no virus.

I use ZoneAlarm in both PCs where I am experiencing the issue.

**Comment by Izheil** [comment]
Seems related to the bugzilla I sent earlier (some other people in there also had ESET antivirus and also had a similar issue... so I assume it's some problem with certain AVs).
 
Change the "lockPref" `true` value to `false`.
It should theorically prevent the locking of more preferences by the AV on Firefox (which seems to cause the issues).

If the `kl_config_62fbb8f7_c917_4cf7_957a_aad2b8fa768c.cfg` file exists in the same folder, change the same lockpref value to "false" as well.

After that see if the "false" values get changed on the next Firefox start, and if they don't, we'll give it a week and see if the problem arises again.

**Comment by simoncarne** [comment]
Thanks. I have made the true/false change you suggested and, so far, it hasn't been reversed. 

So far as I can tell, the `kl_configkkl_config_62fbb8f7_c917_4cf7_957a_aad2b8fa768c.cfg` file doesn't exist anywhere on my PC. 

Let's see how it goes now .

**Comment by simoncarne** [comment]
Thanks so much for your time on this.

**Comment by Izheil** [comment]
No problem! 
I'll leave this open for now for a week or 2, and if you find any issue, report it back and we'll see what can we do about it then.

**Comment by simoncarne** [comment]
I have a bit of an update.
1) The change we made (changing "lockPref" true value to false) seems to be working fine.
2) Although I reported previously that temporarily disabling ZoneAlarm didn't make any difference to the problem, I have re-tested that and turning off ZA does in fact stop the problem occurring. (I must have messed-up first time - apologies.) I have reported the problem to ZA. They are looking into it. I will report back here when I hear more from ZA.
